lib/pokerscore/hand.rb

```diff
@@ -1,83 +1,48 @@
-require './const/hand_constants.rb'
+require 'const/hand_constants.rb'
+require 'lib/pokerscore/set.rb'
 
 class Hand
-  include Comparable
-  attr_reader :cards, :straight_high_card, :sets, :bestcards
+  attr_reader :cards
 
-  def <=>(hand2)
-    if bestcards != hand2.bestcards
-      return $hand_types[bestcards] <=> $hand_types[hand2.bestcards]
-    end
-    if straight_high_card != nil
-      return straight_high_card <=> hand2.straight_high_card 
-    end
-    values = []
-    hand2_values = []
-    (1..4).each { |index| values += sets[index].sort }
-    (1..4).each { |index| hand2_values += hand2.sets[index].sort }
-    values.zip(hand2_values).each do |value, hand2_value|
-      return value <=> hand2_value if value != hand2_value
-    end
-  end
-
-  def to_s 
-    cards.map(&:to_s).join(", ")
-  end
-  
   def initialize(cards)
     raise ArgumentError if invalid_hand?(cards)
     @cards = cards
-    @straight_high_card = find_straight(cards)
-    @sets = find_sets(cards)
-    @bestcards = find_best_cards(cards)
   end
-
-  def invalid_hand?(cards)
-    unique_cards = cards.uniq {|card| [card.suit, card.value]}
-    unique_cards.length != 5
+  
+  def sets
+    sets = []
+    unique_values = cards.map(&:value).uniq
+    unique_values.each do |unique_value|
+      occurrances = cards.count { |card| card.value == unique_value }
+      sets << Set.new(unique_value, occurrances)
+    end
+    sets
   end
-
-  def flush?
-    cards.map(&:suit).uniq.length == 1
+  
+  def high_card
+    cards.max { |card1, card2| card1.value <=> card2.value }.value
   end
-
-  private
-
-
-  def find_straight(cards)
+  
+  def straight?
+    result = true
     sorted_values = cards.map(&:value).sort
-    result = sorted_values.last
     sorted_values.each_cons(2).each do |value, next_value|
-      result = nil if not value == next_value - 1
+      result = false if not value == next_value - 1
     end
-    result = 5 if sorted_values == [2,3,4,5,14] # wheel straight
+    result = true if sorted_values == [2,3,4,5,14] # wheel straight
     result
   end
```

Comment:
We may be able to clean this method up a bit.

One approach that I'm fond of is making early returns. One advantage to doing this, is that you don't always have to parse through the entire method for every control flow. It also reduces the cognitive complexity the further you go, as you know that some cases have already been accounted for and would have returned by now.

For example:

```
def straight?
  sorted_values = cards.map(&:value).sort
  
  # By putting the wheel straight into a constant, we can kill the comment 
  return true if sorted_values == WHEEL_STRAIGHT

  sorted_values.each_cons(2).each do |value, next_value|
    return false if value != next_value - 1
  end

  true
end
```

For an even cleaner look, we could also throw `cards.map(&:value).sort` into its own method 🎉

Reply:
Nice, thanks, I agree. I've cleaned it up but I'm leaving that map/sort as it's own line for now, will deal with it down the line with some other related changes.